crates/cli/src/codegen/mod.rs

```diff
@@ -259,11 +260,35 @@ impl Generator {
             .i32_const(0) // offset into data segment for mem.init
             .i32_const(bytecode_len) // size to copy from data segment
             // top-2: dest addr, top-1: offset into source, top-0: size of memory region in bytes.
-            .memory_init(imports.memory, bytecode_data)
-            // Evaluate bytecode.
-            .local_get(bytecode_ptr_local) // ptr to bytecode
-            .i32_const(bytecode_len)
-            .call(imports.eval_bytecode);
+            .memory_init(imports.memory, bytecode_data);
+        // Evaluate top level scope.
+        if let Some(eval_bytecode) = imports.eval_bytecode {
+            instructions
+                .local_get(bytecode_ptr_local) // ptr to bytecode
+                .i32_const(bytecode_len)
+                .call(eval_bytecode);
+        } else {
+            // Assert we're not emitting a call with a null function to
+            // invoke for `javy_quickjs_provider_v*`.
+            // `javy_quickjs_provider_v2` will never support calling `invoke`
+            // with a null function. Older `javy_quickjs_provider_v3`'s do not
+            // support being called with a null function. User plugins and
+            // newer `javy_quickjs_provider_v3`s do support being called with a
+            // null function.
+            // Using `assert!` instead of `debug_assert!` because integration
+            // tests are executed with Javy built with the release profile so
+            // `debug_assert!`s are stripped out.
+            assert!(
```

Comment:
My plan is to relax this assertion a bit after we have released a new version of `javy_quickjs_provider_v3`.